### PR TITLE
fix(auth): verify Central tokens during bootstrap

### PR DESCRIPTION
**Pull request by tanmoysrt**
## Issue
Pilot can reject a valid Central site-login token with 401 when the request arrives before the bootstrap watcher persists the staged JWKS audience.

## Summary
Use the staged IMDS JWKS issuer only while Central bootstrap is pending. Use the persisted shared configuration after bootstrap completes.

## What changed
- Read the staged JWKS URL, audience, and initial key set during the bootstrap window.
- Keep completed and self-hosted benches on their configured issuer.
- Add regression tests for both sides of the bootstrap boundary.
- Document the authentication handoff.

## Why
This removes the startup race without weakening audience validation or extending IMDS use beyond bootstrap.

## Validation
- uv run ruff check admin/backend/internal/session.py tests/admin/backend/test_jwks.py
- uv run pytest tests/admin/backend/test_jwks.py tests/admin/backend/test_central_bootstrap.py -q

### CODE REVIEW

File: admin/backend/internal/session.py [hunk 1]
```diff
@@ -304,3 +304,26 @@ def _decode_jwks(self, token: str) -> dict | None:
...
+            credentials = InstanceMetadata().get_credentials()
+        except CentralClientError as error:
+            logging.warning("Cannot use the staged Central JWKS issuer: %s", error)
+            return url, audience
+        if credentials is None:
+            return url, audience
+
+        url = credentials["jwks_url"]
+        if initial_cache := credentials.get("initial_jwks_cache"):
+            JwksCache(self.bench.path.parent, url).seed(initial_cache)
```
[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py]: Valid. Fixed in 85d86ba3: JwksCache.seed() now initializes only an empty or unusable cache under the cache file lock, so staged keys cannot replace keys fetched after rotation. Added a regression test that preserves the fetched key set.

File: admin/backend/internal/session.py [hunk 2]
```diff
@@ -158,6 +159,8 @@ class Session:
         "PS512",
         "EdDSA",
     ]
+    _staged_jwks_configs: ClassVar[dict[Path, tuple[str, str]]] = {}
+    _staged_jwks_lock: ClassVar[threading.Lock] = threading.Lock()
```
[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py]: Valid. Fixed in acdb1b7a: removed the process-wide staged issuer cache. While bootstrap is pending, authentication reads the current IMDS issuer only; after bootstrap it reads persisted config only. The IMDS-provided initial key set remains the no-issuer-fetch fast path.

File: tests/admin/backend/test_jwks.py [hunk 3]
```diff
@@ -134,6 +134,50 @@ def test_no_audience_config_rejects_remote_token() -> None:
...
+    }
+    monkeypatch.setattr(
+        "pilot.integrations.central.metadata.InstanceMetadata.get_credentials",
+        lambda self: credentials,
+    )
+
+    claims = Session(bench).verify_token(_mint(aud="vm-boot-1"))
+
+    assert claims and claims["sub"] == "admin"
+    assert JwksCache(tmp_path / "benches", JWKS_URL).signing_key("rsa-key") is not None
```
[the remark this answers is not in the document]
Comment (tanmoysrt) [reply tests/admin/backend/test_jwks.py:174]: Valid. Fixed in acdb1b7a: this test now makes PyJWKClient.fetch_data fail explicitly, so successful verification proves that the IMDS initial_jwks_cache was seeded and used without an issuer fetch.

File: admin/backend/internal/session.py [hunk 4]
```diff
@@ -304,3 +304,26 @@ def _decode_jwks(self, token: str) -> dict | None:
...
+        url = getattr(self.admin_config, "jwks_url", "")
+        audience = getattr(self.admin_config, "jwks_audience", "")
+        central = getattr(self.bench.config, "central", None)
+        if not getattr(central, "is_awaiting_bootstrap", False):
+            return url, audience
+
+        from pilot.integrations.central import CentralClientError, InstanceMetadata
+
+        try:
+            credentials = InstanceMetadata().get_credentials()
```
[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py:366]: But call to instance metadata is very fast, should be fine.

[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py:366]: Unless its hitting central's jwks endpont, it will be fine

[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py:366]: Valid. Fixed in ad597f9a with a lifecycle-aware per-process cache. Pending requests reuse the staged IMDS issuer while common_config.toml has the same generation. Any bootstrap or rebootstrap config write changes that generation and forces a fresh IMDS read; observing completed bootstrap removes the entry. Tests cover reuse, config-change invalidation, completion cleanup, and the no-JWKS-fetch fast path.

File: admin/backend/internal/session.py [hunk 5]
```diff
@@ -304,3 +307,51 @@ def _decode_jwks(self, token: str) -> dict | None:
...
+
+        directory = self.bench.path.parent
+        generation = self._common_config_generation(directory)
+        with self._staged_jwks_lock:
+            cached = self._staged_jwks_configs.get(directory)
+            if cached is not None and cached[0] == generation:
+                return cached[1], cached[2]
+            config = self._load_staged_jwks_config(directory)
+            if config != ("", ""):
+                self._staged_jwks_configs[directory] = (generation, *config)
```
[the remark this answers is not in the document]
Comment (tanmoysrt) [reply admin/backend/internal/session.py]: Valid. Fixed in 03917f46 with a per-generation single-flight entry. The lock now protects only cache lookup/insertion; IMDS I/O and Future.result() both run outside it. Concurrent requests for the same pending bootstrap share one lookup, while empty results are discarded so a later request retries. Added concurrency and retry regression tests; 60 focused tests pass.
